Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/RewardsReporting.js

```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+if (webkit.messageHandlers.rewardsReporting) {
+    install();
+}
+
+function install() {
+    function sendMessage(method, url, data, referrerUrl) {
+        webkit.messageHandlers.rewardsReporting.postMessage({
+            method: method === undefined ? "GET" : method,
+            url: url,
+            data: data === undefined ? null : data,
+            referrerUrl: referrerUrl === undefined ? null : referrerUrl,
+        });
+    }
+    
+    let originalOpen = null;
+    let originalSend = null;
+    let originalImageSrc = null;
+    let mutationObserver = null;
+    let originalFetch = null;
+    let originalSendBeacon = null;
+    
+    var xhrProto = XMLHttpRequest.prototype;
+
+    originalOpen = xhrProto.open;
+    originalSend = xhrProto.send;
+
+    xhrProto.open = function(method, url) {
+        const listener = function() {
+            sendMessage(this._method, this.responseURL === null ? this._url : this.responseURL, this._data, this._ref);
+        };
+        this._method = method;
+        this._url = url;
+        this.addEventListener('load', listener, true);
+        this.addEventListener('error', listener, true);
+        return originalOpen.apply(this, arguments);
+    };
+    
+    xhrProto.send = function(body) {
+        this._ref = null;
+        this._data = body;
+        if (body instanceof Document) {
+            this._ref = body.referrer;
+            this._data = null;
+        }
+        return originalSend.apply(this, arguments);
+    };
+
+    originalFetch = window.fetch;
+    window.fetch = function(resource, options) {
+        const args = arguments
+        const url = resource instanceof Request ? resource.url : resource
+
+        return new Promise(function(resolve, reject) {
+            originalFetch.apply(this, args)
+            .then(function(response) {
+                sendMessage(options.method, url, options.body, options.referrer);
+                resolve(response);
+            })
+            .catch(function(error) {
+                sendMessage(options.method, url, options.body, options.referrer);
+                reject(error);
+            })
+        });
+    };
+
+    originalSendBeacon = navigator.sendBeacon;
+    navigator.sendBeacon = function(url, data) {
+        sendMessage("POST", url, data);
+        return originalSendBeacon.apply(this, arguments);
+    };
+    
+    originalImageSrc = Object.getOwnPropertyDescriptor(Image.prototype, "src");
+    delete Image.prototype.src;
+    Object.defineProperty(Image.prototype, "src", {
```

Comment:
Not sure if what I'm saying is necessary, but `HTMLImageElement` (aka `Image`) has 3 source properties:

```
src
lowsrc
currentSrc
```

lol :(
If the page has a `media` element, it "might" choose a different source based on resolution and device. Does that in any way affect this code?

Reply:
The simple answer to that is: I don't know :) 

It works for the set of publishers at the moment (YouTube specifically uses Image src's), and as far as I know none of the other publishers use it